### PR TITLE
Add support for setting the settings configuration path via CLI and import

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

Add support for loading settings from a different path. Replaces `SETTINGS` by a `get_settings` function.

Related to https://github.com/napari/napari/issues/2336

Specifically it fixes:

>Does not use global settings object.
>Why: When using napari from notebook user may open more than one napari Viewer instance. They should not share state so should not share settings object. Open multiple Viewer may be useful when preparing to meet and like to prepare few different views (without a wait for calculations on meeting).

>?Allow to simply provide a path to the directory where settings should be stored.
>Why: I embed napari in my own application This application provides its own set of plugins etc. To avoid settings collision (especially when store plugins-related things in napari) I would like to store settings for the embedded instance application-specific directory. I think that it also will be useful for other persons who develop rich plugins.

Pinging @Czaki for testing :) 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by tlambert03** [comment]
ah, interesting... I can see how #2662 would be useful here.  Did you have any thoughts on #2662 as a way to get to `__main__` without greedily importing everything?  I'm sure there are some other strategies we could employ... but it seems to get the job done.

**Comment by goanpeca** [comment]
I am all for that PR, now that there is a need for it, I like it even more :-p

**Comment by sofroniewn** [comment]
#2662 has now merged, @goanpeca does this need to be updated?

Also @Czaki can you weigh in on if you like this change. Thanks!!

**Comment by goanpeca** [comment]
>#2662 has now merged, @goanpeca does this need to be updated?

I already did that, but I do need to fix the merge conflict.

**Comment by goanpeca** [comment]
I would also like to hear from @napari/core-devs 

As mentioned originally when we started working with the settings, the idea was to be able to not access a global object directly, now we do this indirectly with this PR but still limits the change to a single session.


So one could do

```bash
napari --settings-path "/some/path/
```

But trying to do `get_settings("/some/other/path")` inside the napari console would error since for the session the config path has been set

Doing something like this would work

```python
from napari.utils.settings import get_settings

settings = get_settings("some-path")
# Do something with napari
# ....
```

But 

```python
from napari.utils.settings import get_settings

settings = get_settings("some-path")
# Do something with napari
# ....

settings = get_settings("some-other-path") # This will error
# Do something else with napari
# .... 
```

So this is the current behavior with the PR and we could update it on another PR (multiviewer case) or what should the behavior be? be able to change settings location when running from scripts but disable when running the application?

I am open for options here :)

**Comment by sofroniewn** [comment]
> As mentioned originally when we started working with the settings, the idea was to be able to not access a global object directly, now we do this indirectly with this PR but still limits the change to a single session.

@goanpeca I like the behaviour described above, but maybe @tlambert03 @jni and @Czaki can weigh in before merge.

File: napari/utils/settings/_manager.py
Comment (Czaki) [inline napari/utils/settings/_manager.py]: Make it a private variable to ensure that it should not be exposed (to prevent some failure in the future).

**Comment by Czaki** [comment]
> > As mentioned originally when we started working with the settings, the idea was to be able to not access a global object directly, now we do this indirectly with this PR but still limits the change to a single session.
> 
> @goanpeca I like the behaviour described above, but maybe @tlambert03 @jni and @Czaki can weigh in before merge.

I have no idea (other than passing settings as argument everywhere) to have multiple settings during one session. I'm happy with this change.

File: napari/utils/settings/_manager.py
Comment (jni) [reply napari/utils/settings/_manager.py]: Well, that would be a breaking change currently. We should do a deprecation cycle, but it is not urgent to start that process.

**Comment by jni** [comment]
Regarding the settings behaviour, in the future, I would very much like a context manager:

```python
with napari.settings('/path/to/settings.yaml'):
    do_the_thing()
```

We could do all kinds of fancy things, e.g. the context manager would not start settings from defaults but from the current settings, then restore them on exit.

And one could even pass in a settings dictionary/object in place of the yaml. In short, it should be similar to the matplotlib styling ux. I'm A-Ok with not supporting that yet, though. @goanpeca do you think that is "easily" feasible?

**Comment by goanpeca** [comment]
Hi @jni, feasible :)

**Comment by Czaki** [comment]
> Regarding the settings behaviour, in the future, I would very much like a context manager:

This will not work with threads. Even determine in which context we are processing pending events in the main thread may be hard.

This is the same problem like determine which viewer instance use to show an error message guessing from which viewer thread was started.

File: napari/utils/settings/_manager.py
Comment (goanpeca) [reply napari/utils/settings/_manager.py]: >Well, that would be a breaking change currently

Yep exactly why it should remain there, it would continue to work the same way.

Comment (Czaki) [reply napari/utils/settings/_manager.py]: But this PR alredy break this.

In every module which uses `from napari.utils.settings._manager import SETTINGS` the value for `SETTINGS` will be `None` 

To have working code it should be from napari.utils.settings import _manager` and called 
`_manager.SETTINGS`

Also in this or `napari/utils/settings/__init__.py` does not export SETTINGS object.

If you would like to kee backwart compatibility there should be a class:


```
class SettingsProxy:
	def __getattribute__(self, name):
		return getattr(get_settings(), name)
```
assigned to `SETTINGS` variable in this module.

**Review by sofroniewn — APPROVED** [review]
Nice, I think this now looks good to go! I'll let others weigh in before a final merge, but I think this steps us in the right direction!!

**Comment by sofroniewn** [comment]
I think this is now good to go, will merge later tonight unless I hear otherwise!!

**Pull request merged**
